lib/tree.rb

```diff
@@ -2,59 +2,124 @@ class TreeNode
   attr_reader :key, :value
   attr_accessor :left, :right
 
-   def initialize(key, val)
+  def initialize(key, val)
     @key = key
     @value = val
     @left = nil
     @right = nil
-   end
+  end
 end
 
 class Tree
   attr_reader :root
+
   def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: o(n)
+  # Space Complexity: o(n)
   def add(key, value)
-    raise NotImplementedError
+    current_node = @root
+
+    if current_node == nil
+      current_node = TreeNode.new(key, value)
+      @root = current_node
+    end
+
+    while nil != current_node
+      if key < current_node.key && current_node.left == nil
+        current_node.left = TreeNode.new(key, value)
+      elsif key > current_node.key && current_node.right == nil
+        current_node.right = TreeNode.new(key, value)
+      elsif key > current_node.key
+        current_node = current_node.right
+      elsif key < current_node.key
+        current_node = current_node.left
+      else
+        return
+      end
+    end
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: o(n)
+  # Space Complexity: 0(1)
   def find(key)
-    raise NotImplementedError
+    return nil if @root.nil?
+    current_node = @root
+
+    return current_node.value if key == current_node.key
+
+    while current_node != key
+      if key == current_node.key
+        return current_node.value
+      elsif key > current_node.key
+        current_node = current_node.right
+      elsif key < current_node.key
+        current_node = current_node.left
+      end
+    end
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+# Time Complexity: O(n)
+# Space Complexity: O(1)
+  # left - Root - Rigth
   def inorder
```

Comment:
👍 , However your space complexity is O(n) because you're building an array of `n` nodes.  Even if you weren't your space complexity would be O(n) or O(log n) because of the system stack.